plugins/core/hu.bme.mit.gamma.action.model/src/hu/bme/mit/gamma/action/util/ActionModelValidator.java

```diff
@@ -174,14 +185,85 @@ protected ActionModelValidator() {}
 	
 	public Collection<ValidationResultMessage> checkBranch(Branch branch) {
 		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
-		// Block is empty
 		Expression guard = branch.getGuard();
-		if (!typeDeterminator.isBoolean(guard)) {
+		EObject container = branch.eContainer();
+		// check if container is a SwitchStatement
+		if (container instanceof SwitchStatement) {
+			SwitchStatement switchStatement = (SwitchStatement) container;
+			if (!typeDeterminator.equalsType(switchStatement.getControlExpression(), guard) && !(branch.getGuard() instanceof DefaultExpression)) {
+				validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+						"Type of control expression must be same type of guard in a SwitchStatement!",
+						new ReferenceInfo(ActionModelPackage.Literals.BRANCH__GUARD)));
+			}
+		}
+		else {
+			if (!typeDeterminator.isBoolean(guard)) {
+				validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+						"Branch conditions must be of type boolean!",
+						new ReferenceInfo(ActionModelPackage.Literals.BRANCH__GUARD)));
+			}
+		}
+		return validationResultMessages;
+	}
+	
+	public Collection<ValidationResultMessage> checkForStatement(ForStatement forStatement) {
+		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
+		// parameter check
+		if (!typeDeterminator.isInteger(forStatement.getParameter().getType())) {
+			validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+					"The type of parameter must be integer!",
+					new ReferenceInfo(ActionModelPackage.Literals.FOR_STATEMENT__PARAMETER)));
+		}
+		// range check 
+		if (!(typeDeterminator.getType(forStatement.getRange()) instanceof IntegerRangeTypeDefinition)) {
 			validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
-					"Brach conditions must be of type boolean",
-					new ReferenceInfo(ActionModelPackage.Literals.BRANCH__GUARD)));
+					"The type of range must be integer range!",
+					new ReferenceInfo(ActionModelPackage.Literals.FOR_STATEMENT__RANGE)));
 		}
 		return validationResultMessages;
 	}
 	
+	public Collection<ValidationResultMessage> checkSwitchStatement(SwitchStatement switchStatement) {
+		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
+		// if controlExpression is an enum, no need the DefaultBranch
+		Type typeDef = typeDeterminator.getTypeDefinition(switchStatement.getControlExpression());
+		if (typeDef instanceof EnumerationTypeDefinition) {
+			List<EnumerationLiteralDefinition> cloneLiterals = ecoreUtil.clone(((EnumerationTypeDefinition) typeDef).getLiterals());
```

Comment:
Wouldn't it be easier just to create a new list with the literals instead of cloning all of them?